src/SemaphoreMultiAirdrop.sol

```diff
@@ -0,0 +1,163 @@
+
+    /// @dev The Semaphore instance that will be used for managing groups and verifying proofs
+    ISemaphore internal immutable semaphore;
+
+    /// @dev Whether a nullifier hash has been used already. Used to prevent double-signaling
+    mapping(uint256 => bool) internal nullifierHashes;
+
+    uint256 internal nextAirdropId = 1;
+    mapping(uint256 => Airdrop) public getAirdrop;
+
+    ///////////////////////////////////////////////////////////////////////////////
+    ///                               CONSTRUCTOR                              ///
+    //////////////////////////////////////////////////////////////////////////////
+
+    /// @notice Deploys a SemaphoreAirdrop instance
+    /// @param _semaphore The Semaphore instance that will manage groups and verify proofs
+    constructor(
+        ISemaphore _semaphore
+    ) {
+        semaphore = _semaphore;
+    }
+
+    /// @notice Create a new airdrop
+    /// @param groupId The ID of the Semaphore group that will be eligible to claim this airdrop
+    /// @param token The ERC20 token that will be airdropped to eligible participants
+    /// @param holder The address holding the tokens that will be airdropped
+    /// @param amount The amount of tokens that each participant will receive upon claiming
+    function createAirdrop(
+        uint256 groupId,
+        ERC20 token,
+        address holder,
+        uint256 amount
+    ) public {
+        Airdrop memory airdrop = Airdrop({
+            groupId: groupId,
+            token: token,
+            manager: msg.sender,
+            holder: holder,
+            amount: amount
+        });
+
+        getAirdrop[nextAirdropId] = airdrop;
+        emit AirdropCreated(nextAirdropId++, airdrop);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////
+    ///                               CLAIM LOGIC                               ///
+    //////////////////////////////////////////////////////////////////////////////
+
+    /// @notice Claim a given airdrop
+    /// @param airdropId The id of the airdrop getting claimed
+    /// @param receiver The address that will receive the tokens
+    /// @param root The of the Merkle tree
+    /// @param nullifierHash The nullifier for this proof, preventing double signaling
+    /// @param proof The zero knowledge proof that demostrates the claimer is part of the Semaphore group
+    function claim(
+        uint256 airdropId,
+        address receiver,
+        uint256 root,
+        uint256 nullifierHash,
+        uint256[8] calldata proof
+    ) public {
+        if (nullifierHashes[nullifierHash]) revert InvalidNullifier();
+
+        Airdrop memory airdrop = getAirdrop[airdropId];
+        if (airdropId == 0 || airdropId >= nextAirdropId) revert InvalidAirdrop();
+
+        semaphore.verifyProof(
+            root,
+            airdrop.groupId,
+            abi.encode(receiver).hashToField(),
```

Comment:
let's make this an abi.encodePacked so we don't need to bother the client

Reply:
same in `SemaphoreAirdrop`

Reply:
We can fix this in a separate PR, it's a bit out of scope for this one.